### DISCUSSION TITLE
Instrument family with different number of channels.

### DISCUSSION THREAD

**Discussion by tomverbeure** [pymeasure/pymeasure#776] [Q&A]
I want to have a single Instrument class that covers a whole family. Different family members have a different number of channels. I'd like to use the ChannelCreator for this, but it looks like this is using class variables to store the channels?

For example, let's look at the AWG401x as an example: https://github.com/pymeasure/pymeasure/blob/8004ad9b05f7e9634b25ea66d4c01bef7f09c90a/pymeasure/instruments/activetechnologies/AWG401x.py#L432-L459

Depending on the model, it defines 2, 4, or 8 channels, which gets assigned to the `ch` class variable.

How does this work when you have a setup with, say, 2 AWG401x_AFG instruments, one with 2 channels and one with 8 channels?

Is there an already implemented instrument that shows how to do this right?

**Comment by neal-kepler** [comment]
I'm not sure this is universal, but that particular class of instruments will tell you how many channels it has and the query string is `AWGC:CONF:CNUM?`

Presumably an attribute like: 

```
num_channels = Instrument.measurement(
        "AWGC:CONF:CNUM?", """Returns the number of analog channels on the instrument"""
    )
```
would be valuable to add. That way you don't even need the `if`'s you can just call `self.num_channels`. This is currently how I solve this exact problem in a fork.

I am not sure if this is a general feature of other AWG remote interfaces, but I don't recall seeing it in any other AWG type

**Reply by tomverbeure** [reply]
Right. But my question in completely different: it's one about channels being stored in a class variable and not an instance variable. (See line 432 in my code snippet above.) ChannelCreator only allows class variables... or, at least, it refuses to work when I use instance variables in my code.

So what I don't understand is how this is supposed to work when you have multiple instances of the same instrument class: since the class variable is shared between multiple instances, they trample onto each other.

There are some other instruments that use the Channel class, but they don't use ChannelCreator, and simply assign to an instance variable.

E.g. the Keithley2306: https://github.com/pymeasure/pymeasure/blob/8004ad9b05f7e9634b25ea66d4c01bef7f09c90a/pymeasure/instruments/keithley/keithley2306.py#L611-L616

But I'm trying to see if I can use the common base classes that are already provided by pymeasure.

**Reply by neal-kepler** [reply]
Ah, you're totally right, my apologies, I was not up to date on the new ChannelCreator, I see you're aim now.

**Comment by BenediktBurger** [comment]
The channels are a very recent addition, so barely any instrument uses them yet. The Keithley uses channels, but it's own implementation (legacy code).

Yes, the Channel Creator works only at class level.
If there are similar instruments (with different code), you can subclass an instrument and override the ChannelCreator.

If the same instrument class can talk with different instruments with different channel numbers, you have to resort to the more basic method `self.add_child`. Parameters are the same (with an additional  `collection`), but it can add only a single channel at a time.
The ChannelCreator calls this function as well.
It is complemented by `remove_child`.
You might want to look in the [documentation](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#instruments-with-channels) for some examples and usage hints.

That way you can dynamically change the channel number (at startup or even later).

If you have questions, don't hesitate to ask.
It is a new feature and maybe we have to improve it or the documentation. Thanks for your feedback.

**Reply by BenediktBurger** [reply]
See the AFG init at https://github.com/pymeasure/pymeasure/blob/8004ad9b05f7e9634b25ea66d4c01bef7f09c90a/pymeasure/instruments/activetechnologies/AWG401x.py#L446-L459

There more channels are added according to the model version.

**Reply by tomverbeure** [reply]
The AFG init adds more channels according to the model, but if I understand the code correctly (not likely!), it's still expanding the `ch` class variable, not an instance variable? So if you have an AWG4012 and an AWG4018, channels 2 to 7 will now be defined even for the AWG4012 instance.

In this particular case, that might still work OK because all channels have the same Channel subclass, but what for devices where, say, channel 0 needs a different subclass for different models?

I haven't tested it, but I wonder if the AWG code works at all when there are multiple instruments: there is currently no check how many channels have already been defined, and it keeps on calling `add_child`, but with duplicate channel numbers. 
So there might be undefined behavior when there's multiple ones.

I don't see any examples in the docs where instance variables are used instead of class variables.

**Reply by BenediktBurger** [reply]
Thanks for the question.

The `add_child` method sets an instance attribute with the name `ch`, which contains the channels. The class attribute `ch` contains still the ChannelCreator object (which is necessary to use more than one instrument of the same class).

I just wrote a test to verify that behaviour.

TLDR: The initial definition (ChannelCreator) is at class level, the channels live at instance level.

**Reply by msmttchr** [reply]
> The AFG init adds more channels according to the model, but if I understand the code correctly (not likely!), it's still expanding the `ch` class variable, not an instance variable? So if you have an AWG4012 and an AWG4018, channels 2 to 7 will now be defined even for the AWG4012 instance.
> 
> In this particular case, that might still work OK because all channels have the same Channel subclass, but what for devices where, say, channel 0 needs a different subclass for different models?
> 
> I haven't tested it, but I wonder if the AWG code works at all when there are multiple instruments: there is currently no check how many channels have already been defined, and it keeps on calling `add_child`, but with duplicate channel numbers. So there might be undefined behavior when there's multiple ones.

@tomverbeure, it is important to understand that `self.ch` is an instance variable and not a class variable.
Also, when using the definition at class level, these are copied behind the scene at instance level during instantiation.
My personal preference is to use a definition at class level as follows for the case you described:

```
class Inst2ChannelModel(Instrument):
     channel = Instrument.ChannelCreator(MyChannel, (1, 2))
    # remaining of implementation
.
.
.
class Inst4ChannelModel(Inst2ChannelModel):
     channel = Instrument.ChannelCreator(MyChannel, (1, 2, 3, 4))

class Inst8ChannelModel(Inst2ChannelModel):
     channel = Instrument.ChannelCreator(MyChannel, (1, 2, 3, 4, 5, 6, 7, 8))

```

**Reply by tomverbeure** [reply]
Thanks!

I was trying to see what it would take to create a generic Agilent power supply instrument. There are probably a hundred different types, and the idea was to have 1 Instrument class with a lookup table based on the `*IDN?*` ID, just like the AWG instrument.

After looking at the user manuals of a couple of PSUs, it's clear that they often also have minor implementation differences.

I'll check it out some more, but there'll probably be more Channel related questions.

**Reply by msmttchr** [reply]
Again, my opinion: I like solution that avoid to be too 'clever', i.e getting id and having features defined at runtime. This sometime is unavoidable, i.e in case of optional features of an instrument, but for the rest I like the approach "one instrument model one python class".
It's also possible, at higher level, to define a table that, based on id result, returns a class to handle the specific instrument.
